ARROW-14511: [Website][Rust] Rust 6.0.0 release blog post (#156)
apache/arrow-site: _posts/2021-11-02-6.0.0-rs-release.md

---
layout: post

> title: Apache Arrow Rust 5.0.0 Release

Author: Thanks

---
layout: post
title: Apache Arrow Rust 6.0.0 Release

> date: "2021-11-08 00:00:00"

Reviewer: ```suggestion
date: "2021-11-09 00:00:00"
```

> author: pmc and contributors

Reviewer: I don't think this works right; either just say 'pmc', or a single person named in https://github.com/apache/arrow-site/blob/master/_data/contributors.yml (possibly adding yourself there), or plain text. (As it currently stands, this would literally print "pmc and contributors" as the author, and I don't think we want this.)

Author: @nealrichardson Leave as PMC. Will add myself to contributors later. So if ok, please merge.

Author: May want to have something in addition to PMC, since a lot of this is/can be automated so project newcomers and committers can also help with this. Maybe there should be PC - Project Contributors? For posts on releases, the role is more of editing and curating rather than authoring despite the suggestion at https://github.com/apache/arrow-site/blob/master/README.md

---
layout: post
title: Apache Arrow Rust 6.0.0 Release

> date: "2021-11-02 00:00:00"

Reviewer: ```suggestion
date: "2021-11-08 00:00:00"
```

author: pmc
categories: [release]
---
<!--
{% comment %}
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

> We recently released the 6.0.0 Rust version of [Apache Arrow](https://arrow.apache.org/) which 

Reviewer: ```suggestion
We recently released the 6.0.0 Rust version of Apache Arrow, which 
```

> coincides with the [Arrow 6.0.0 release](https://arrow.apache.org/release/6.0.0.html). This 

Reviewer: ```suggestion
coincides with the [Arrow 6.0.0 release]({{ site.baseurl }}/6.0.0.html). This 
```

post highlights some of the improvements in the Rust implementation. The full changelog can be 
found [here](https://github.com/apache/arrow-rs/blob/6.0.0/CHANGELOG.md).

<!--
(arrow_dev) bkmgit@Linux:~/arrow-rs$ git log --pretty=oneline 5.0.0..6.0.0 | wc -l
     99
(arrow_dev) bkmgit@Linux:~/arrow-rs$ git shortlog -sn 5.0.0..6.0.0 | wc -l
     35
-->

The Rust Arrow implementation would not be possible without the wonderful work and support of our community, and 
the 6.0.0 release is no exception. It includes 99 commits from 35 individual contributors, many of them with 
their first contribution. Thank you all very much.

# Arrow

Highlighted features and changes between release 5.0.0 and this release are:
1. New MapArray support
2. Add optimized filter kernel for regular expression matching

> 3. Support arrow map datatype

Reviewer 2: ```suggestion
```

I think this is redundant with the first item

> 4. Implement sort() for binary array

Reviewer 2: ```suggestion
4. Implement `sort()` for `BinaryArray`
```

5. Replace `ArrayData::new()` with `ArrayData::try_new()` and `unsafe ArrayData::new_unchecked`
6. Sorting should require less memory and be faster

> We continue to leverage the Rust ecosystem to deliver reliable and performant code.

Reviewer: I don't really understand what this sentence means

Author: Sentence `We continue to leverage the Rust ecosystem to deliver reliable and performant code.` removed

Of course, this release also contains bug fixes, performance improvements, and improved documentation examples. For the full list of changes, 
please consult the [changelog](https://github.com/apache/arrow-rs/blob/6.0.0/CHANGELOG.md).

# More Frequent Releases

> Arrow releases major versions every three months. The Rust implementation has been experimenting with releasing minor version updates to 
> speed the flow of new features and fixes.

Reviewer 2: ```suggestion
Arrow releases major versions every three months. The Rust implementation follows this 
major release cycle, and additionally releases minor version updates approximately every other week
to speed the flow of new features and fixes.
```

You can always find the latest releases on crates.io: [`arrow`](https://crates.io/crates/arrow), [`parquet`](https://crates.io/crates/parquet), 
[`arrow-flight`](https://crates.io/crates/arrow-flight), and [`parquet-derive`](https://crates.io/crates/parquet-derive).

# DataFusion & Ballista

> [DataFusion](https://docs.rs/datafusion/4.0.0/datafusion/) is an in-memory query engine with DataFrame and SQL APIs, built on top of Arrow. 

Reviewer: ```suggestion
[DataFusion](https://docs.rs/datafusion/) is an in-memory query engine with DataFrame and SQL APIs, built on top of Arrow. 
```

Reviewer 2: How about using this doc link: https://arrow.apache.org/datafusion/

Ballista is a distributed compute platform. These projects are now in their [own repository](https://github.com/apache/arrow-datafusion), 
and are no longer released in lock-step with Arrow.

# Highlighted Functionality

> The memory required to do sorting has been improve by pull request resolvin issue [553](https://github.com/apache/arrow-rs/issues/553). 

Reviewer: ```suggestion
The memory required to do sorting has been improve by pull request resolving issue [553](https://github.com/apache/arrow-rs/issues/553). 
```

Reviewer 2: ```suggestion
The memory required to do sorting has been improve by the pull request resolving issue [553](https://github.com/apache/arrow-rs/issues/553). 
```

A demonstration for how to sort follows.
```rust
extern crate arrow;

use arrow::array::{
    Int32Array,
    ArrayRef,
};
use std::sync::Arc;
use arrow::compute::sort;

fn main() {
    
    let array: ArrayRef = Arc::new(Int32Array::from(vec![5, 4, 23, 1, 20, 2]));
    println!("{:?}", array);
    let sorted_array = sort(&array, None).unwrap();
    println!("{:?}", sorted_array);
}
```
For further examples, see the [source code](https://github.com/apache/arrow-rs/blob/master/arrow/src/compute/kernels/sort.rs).

# Roadmap for 7.0.0 and Beyond
Here are some of the initiatives that contributors are currently working on for future releases:

* Validate arguments to ArrayData::new and null bit buffer and buffers
* add ilike comparitor
* refactor regexp_is_match_utf8_scalar
* add support for float 16

> 

Reviewer 2: ```suggestion
* Updated `UnionArray` support to follow the latest arrow spec

```
